File: threads/src/main/java/im/threads/internal/database/table/Table.java
package im.threads.internal.database.table;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

public abstract class Table {

    public abstract void createTable(SQLiteDatabase db);

    public abstract void upgradeTable(SQLiteDatabase db, int oldVersion, int newVersion);

    public abstract void cleanTable(SQLiteOpenHelper sqlHelper);

    static boolean cIsNull(Cursor c, String columnName) {
        return c.isNull(c.getColumnIndex(columnName));
    }

    static boolean cGetBool(Cursor c, String columnName) {
        return cGetInt(c, columnName) == 1;
    }

    @Nullable
    static String cGetString(Cursor c, String columnName) {
        return cIsNull(c, columnName) ? null : c.getString(c.getColumnIndex(columnName));
    }

    static long cGetLong(Cursor c, String columnName) {
        return c.getLong(c.getColumnIndex(columnName));
    }

    static int cGetInt(Cursor c, String columnName) {
        return cIsNull(c, columnName) ? 0 : c.getInt(c.getColumnIndex(columnName));
    }

}

File: threads/src/main/java/im/threads/business/database/table/Table.kt
package im.threads.business.database.table

import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

abstract class Table {
    abstract fun createTable(db: SQLiteDatabase)
    abstract fun upgradeTable(db: SQLiteDatabase, oldVersion: Int, newVersion: Int)
    abstract fun cleanTable(sqlHelper: SQLiteOpenHelper)

    companion object {
        fun cIsNull(c: Cursor, columnName: String?): Boolean {
            val index = c.getColumnIndex(columnName)
            return index < 0 || c.isNull(index)
        }

        fun cGetBool(c: Cursor, columnName: String?): Boolean {
            return cGetInt(c, columnName) == 1
        }

        fun cGetString(c: Cursor, columnName: String?): String? {
            val index = c.getColumnIndex(columnName)
            return if (index < 0 || cIsNull(c, columnName)) {
                null
            } else {
                c.getString(index)
            }
        }

        fun cGetLong(c: Cursor, columnName: String?): Long {
            val index = c.getColumnIndex(columnName)
            return if (index < 0) {
                0
            } else {
                c.getLong(index)
            }
        }

        fun cGetInt(c: Cursor, columnName: String?): Int {
            val index = c.getColumnIndex(columnName)
            return if (index < 0 || cIsNull(c, columnName)) {
                0
            } else {
                c.getInt(index)
            }
        }
    }
}
